### PR TITLE
[16.0][ADD] stock_release_channel_delivery

### PR DESCRIPTION
**Pull request by glitchov**
This addon adds a field containing the carrier linked to the release channel. 
Though an overlap can be seen on the picking point of view as the
rule_domain does the job the idea is to easy link the channel with the
customer.

### CODE REVIEW

**Review by jbaudoux — APPROVED** [review]

**Comment by jbaudoux** [comment]
@sbejaoui Can you rebase, it seems failing tests are not related to this PR

**Comment by sbejaoui** [comment]
> @sbejaoui Can you rebase, it seems failing tests are not related to this PR

done

File: stock_release_channel_delivery/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Copyright 2023 ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Stock Release Channel Delivery",
+    "summary": """
+        Exposes clearly the carrier linked to the release channel """,
```
Comment (jbaudoux) [inline stock_release_channel_delivery/__manifest__.py]:
```suggestion
        Add a carrier selection criteria on the release channel """,
```

File: stock_release_channel_delivery/models/stock_picking.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+from odoo import models
+
+
+class StockPicking(models.Model):
+
+    _inherit = "stock.picking"
+
+    def _get_release_channel_possible_candidate_domain(self):
+        domain = super()._get_release_channel_possible_candidate_domain()
+        if self.carrier_id:
```
Comment (jbaudoux) [inline stock_release_channel_delivery/models/stock_picking.py:13]:
```suggestion
        if self.carrier_ids:
```

File: stock_release_channel_delivery/models/stock_picking.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+            domain.extend(
+                [
+                    "|",
+                    ("carrier_id", "=", False),
+                    ("carrier_id", "=", self.carrier_id.id),
```
Comment (jbaudoux) [inline stock_release_channel_delivery/models/stock_picking.py:18]:
```suggestion
                    ("carrier_ids", "=", False),
                    ("carrier_ids", "in", self.carrier_id.ids),
```

File: stock_release_channel_delivery/models/stock_release_channel.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class StockReleaseChannel(models.Model):
+
+    _inherit = "stock.release.channel"
+
+    carrier_id = fields.Many2one(
```
Comment (jbaudoux) [inline stock_release_channel_delivery/models/stock_release_channel.py:11]:
```suggestion
    carrier_ids = fields.Many2many(
```

File: stock_release_channel_delivery/models/stock_release_channel.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+        comodel_name="delivery.carrier",
+        help="The carrier linked to the release channel.",
+        string="Transporter",
```
Comment (jbaudoux) [inline stock_release_channel_delivery/models/stock_release_channel.py]:
```suggestion
        string="Shipping Methods",
```

File: stock_release_channel_delivery/readme/DESCRIPTION.rst [hunk 6]
```diff
@@ -0,0 +1,4 @@
+This addon adds a field containing the id of the carrier linked to the release
+channel. Though an overlap can be seen on the picking point of view as the
+rule_domain does the job the idea is to easy link the channel with the
+customer.
```
Comment (jbaudoux) [inline stock_release_channel_delivery/readme/DESCRIPTION.rst]:
```suggestion
This addon adds a selection criteria on the release channel based on the shipping method.
When you don't classify your deliveries in different operation types per shipping method, then you can use this criteria to filter the selection of valid release channels. While you could achieve the same with the domain provided on the base release channel module, this criteria provides a more explicit configuration and also performs better.
```

File: stock_release_channel_delivery/views/stock_release_channel_views.xml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+            name="name"
+        >stock.release.channel.form (in stock_release_channel_delivery)</field>
+        <field name="model">stock.release.channel</field>
+        <field
+            name="inherit_id"
+            ref="stock_release_channel.stock_release_channel_form_view"
+        />
+        <field name="arch" type="xml">
+            <field name="rule_domain" position="before">
+                <field name="carrier_id" />
```
Comment (jbaudoux) [inline stock_release_channel_delivery/views/stock_release_channel_views.xml:15]:
```suggestion
                <field name="carrier_ids" widget="many2many_tags"/>
```

File: stock_release_channel_delivery/tests/test_release_channel_delivery.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        cls.channel = cls.env.ref("stock_release_channel.stock_release_channel_default")
+        cls.move = cls._create_single_move(cls.product1, 10)
+        cls.picking = cls.move.picking_id
+
+    def test_00(self):
+        """
+        picking and release channel have the same carrier
+        """
+        self.picking.carrier_id = self.carrier
+        self.channel.carrier_id = self.carrier
```
Comment (jbaudoux) [inline stock_release_channel_delivery/tests/test_release_channel_delivery.py:23]:
```suggestion
        self.channel.carrier_ids = self.carrier
```

File: stock_release_channel_delivery/tests/test_release_channel_delivery.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        self.picking.assign_release_channel()
+        self.assertEqual(self.picking.release_channel_id, self.channel)
+
+    def test_01(self):
+        """
+        picking have carrier but not the release channel
+        """
+        self.picking.carrier_id = self.carrier
+        self.channel.carrier_id = False
```
Comment (jbaudoux) [inline stock_release_channel_delivery/tests/test_release_channel_delivery.py:32]:
```suggestion
        self.channel.carrier_ids = False
```

File: stock_release_channel_delivery/tests/test_release_channel_delivery.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        self.picking.assign_release_channel()
+        self.assertEqual(self.picking.release_channel_id, self.channel)
+
+    def test_02(self):
+        """
+        picking and release channel have different carrier
+        """
+        self.picking.carrier_id = self.carrier
+        self.channel.carrier_id = self.carrier2
```
Comment (jbaudoux) [inline stock_release_channel_delivery/tests/test_release_channel_delivery.py:41]:
```suggestion
        self.channel.carrier_ids = self.carrier2
```

File: stock_release_channel_delivery/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    "version": "16.0.1.0.1",
+    "license": "AGPL-3",
+    "author": "ACSONE SA/NV,Odoo Community Association (OCA)",
```
Comment (jbaudoux) [inline stock_release_channel_delivery/__manifest__.py]:
```suggestion
    "author": "ACSONE SA/NV,BCIM,Odoo Community Association (OCA)",
```

File: stock_release_channel_delivery/readme/CONTRIBUTORS.rst [hunk 12]
```diff
@@ -0,0 +1 @@
+* Hughes Damry <hughes.damry@acsone.eu>
```
Comment (jbaudoux) [inline stock_release_channel_delivery/readme/CONTRIBUTORS.rst:1]:
```suggestion
* Hughes Damry <hughes.damry@acsone.eu>
* Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
```

**Review by jbaudoux — CHANGES_REQUESTED** [review]
We cannot be so restrictive to only one shipping method on the release channel.

In odoo, there are some providers requiring to create one shipping method for national deliveries and another one for international deliveries. However they are usually managed the same way in the warehouse and classified on the same release channel.

**Comment by simahawk** [comment]
@glitchov are you planning to move fwd soon here? Otherwise we could take over.

**Comment by simahawk** [comment]
ping :)

**Comment by lmignon** [comment]
@simahawk We are stuck on stabilizing our Odoo 16 instance.... We've no time to manage the change and the migration of a `Many2one` to `Many2many` field type for the carrier on the release channel model.

**Comment by jbaudoux** [comment]
replaced by https://github.com/OCA/wms/pull/840

**Pull request closed without merging**